Replace github library with graphql

annajung opened the pull request:
Signed-off-by: Anna Jung (VMware) <antheaj@vmware.com>

- Replace the use of GitHub library with graphql 
- Add unit test for `githubDataExtraction.py`

This PR is the first step to fixing https://github.com/vmware-labs/ml-conversational-analytic-tool/issues/3. Replacing GitHub library with graphql helps to gather more data in less time, but the issue of GitHub rate limit still remains for large repos due to [500,000 node limitation](https://docs.github.com/en/graphql/overview/resource-limitations#node-limit) and [rate limit of 5,000 points per hour](https://docs.github.com/en/graphql/overview/resource-limitations#rate-limit).

Based on the [resource limitations doc](https://docs.github.com/en/graphql/overview/resource-limitations), if we are pulling data for N number of repo, then the total nodes used per call calculates should be similar to below:
- Pull requests: N * 100 = N * 100.
- Comments: N * 100 * 50 = N * 5000
- Reviews: N * 100 * 50 = N * 5000
- Reactions:  N * 100 * 50 = N * 5000 

For one repo, the total nodes would be >=  15,100 per call
For 10 repos, total nodes would be >=  151,000 per call 
(Note: number for pull request (100), number for comments (50), number for reviews (50), number for reactions (50) can be configured. accepted values are 1-100.)

With pagination, we are only able to collect up to 100 pull requests for a call. Meaning, if the repo has more than 100 pull requests, it would make the request again until the rate limit is hit. 

The GraphQL API v4 rate limit is 5,000 points per hour. If I understand the doc correctly, points are calculated by dividing up nodes by 100 (15,100 / 100) which is ~ 151 points per call for one repo.

Replacing it with graphql does increase the performance and reduces the number of requests made. 
I believe it does help resolve the issue, but to completely fix the issue I think we should look at providing options to store the page cursor and try again once the limit is reset or an option to use multiple keys (for teams).

difince reviewed (changes requested):
Thank you @anna
One small nit, not directly related to your code 
Running a specific test as described in the README file fails with: 

> ImportError: Failed to import test module: test_githubDataExtraction
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
ModuleNotFoundError: No module named 'tests.test_githubDataExtraction' 

Let make /test a package

sechkova commented on ml-conversational-analytic-tool/githubDataExtraction.py:
```diff
@@ -157,23 +211,21 @@ def getPullFeatures(self, pull):
         # Extract data for an individual repository
-        extractor.openRepo(args.organization, args.repo)
-        df = extractor.getAllPulls(args.reactions)
+        df = extractor.get_all_pull_requests(args.repo, args.reactions)
```
These three lines are repeated for each repository. Could go into a separate function I think.

annajung replied:
refactored to create a separate function

sechkova commented on ml-conversational-analytic-tool/githubDataExtraction.py:
```diff
@@ -1,154 +1,208 @@
+        pull_request_df = pd.DataFrame(pull_data)
+
+        try:
+            return pull_request_df.sort_values(by=pull_request_df.columns[0])
```
nit: Can you add a comment mentioning the name of the field by which the values are sorted?

annajung replied:
Done

sechkova commented on ml-conversational-analytic-tool/tests/test_githubDataExtraction.py:
```diff
@@ -0,0 +1,483 @@
+        self.assertEqual([], actual)
+
+    def test_get_all_pull_requests_returns_dataframe(self):
+        raw_data_file = open("{}/resources/pr_raw_data_example_one.json".format(self.directory_path))
```
The tests should be using "with" statements otherwise the file remains unclosed. Same for all the tests below.

annajung replied:
Thanks for catching this! all fixed PTAL

sechkova commented on ml-conversational-analytic-tool/githubDataExtraction.py:
```diff
@@ -157,23 +211,21 @@ def getPullFeatures(self, pull):
     parser.add_argument('--reactions', action='store_true', default=False, help='Flag to extract reactions')
     parser.add_argument('-n', "--name", help='Output file name. If not specified, the name is constructed like this: '
-            '<organization>_<repo>.csv')
+                                             '<organization>_<repo>.csv')
```
nit: seems like an unintended tabulation?

sechkova reviewed (commented):
@annajung I got curious about the graphql API and had a look. Besides the performance benefits you've mentioned, the API seems to be more flexible too, it looks like a good improvement.

There is still a dependency on PyGithub that should be removed now? https://github.com/vmware-labs/ml-conversational-analytic-tool/blob/5255f259e74fd163c8aa835109b0752141da422f/requirements.txt#L3

I also left some minor questions and comment and noticed warnings about unclosed resources in the CI.

sechkova reviewed (approved):
Thanks, @annajung

difince reviewed (approved):
Thank you, Anna 
LGTM

ericwb commented on ml-conversational-analytic-tool/queries/pull_requests_with_reactions.graphql:
```diff
@@ -0,0 +1,68 @@
+            }
+        }
+    }
+}
```
```suggestion
}

```

ericwb commented on ml-conversational-analytic-tool/queries/pull_requests_without_reactions.graphql:
```diff
@@ -0,0 +1,59 @@
+            }
+        }
+    }
+}
```
```suggestion
}

```

ericwb commented on ml-conversational-analytic-tool/queries/repos.graphql:
```diff
@@ -0,0 +1,16 @@
+            }
+        }
+    }
+}
```
```suggestion
}

```

ericwb commented on ml-conversational-analytic-tool/utils.py:
```diff
@@ -38,7 +38,4 @@ def construct_file_name(name, component_a, component_b, separator="_"):
-    for i in range(len(string)):
-        string[i] = ast.literal_eval(string[i])
-    return string
+    return ast.literal_eval(string)
```
```suggestion
    return ast.literal_eval(string)

```

ericwb reviewed (commented):
Some nits

annajung commented:
I made changes based on @ericwb 's feedback and incorporated waiting when the rate limit has exceeded. Please take another look.

With the new change, I was able to pull data from a repo with ~4500 pull requests in ~15mins without exiting when rate limit exceeds.

The only support that is missing is pagination support for comments, reviews, and reactions when it exceeds 100. I think it can be a separate PR, but let me know if it should be included here.

ericwb commented on ml-conversational-analytic-tool/tests/__init__.py:
```diff
@@ -0,0 +1,2 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
```
```suggestion
# SPDX-License-Identifier: Apache-2.0

```

difince commented:
It figured out that there are two types of Github Rate limit 
1. Limits the number of requests per hour (see [here](https://docs.github.com/en/rest/overview/resources-in-the-rest-api#rate-limiting)) 
If we hit this error - strangely but the response status is == 200 and the entire message looks like this: 

  ```{
  "errors": [
    {
      "type": "RATE_LIMITED",
      "message": "API rate limit exceeded for user ID 11557050."
    }
  ]
}
```


  It is missing the `data` and because the status ==200 there is an exception:
```

  Traceback (most recent call last):
  File "/home/didi/git/anna/ml-conversational-analytic-tool/ml-conversational-analytic-tool/githubDataExtraction.py", line 105, in run_queries
    repo_pull_requests = response_json['data']['repository']['pullRequests']
KeyError: 'data'

```


2. the second type of rate limit is "[Secondary rate limits](https://docs.github.com/en/rest/overview/resources-in-the-rest-api#secondary-rate-limits)". It is an additional rate limit that may be hit in cases when we sent too many requests in a short period of time. After a few seconds, the request will be successfully executed.  It has status code 403 

```
 HTTP/2 403
> Content-Type: application/json; charset=utf-8
> Connection: close

> {
>   "message": "You have exceeded a secondary rate limit and have been temporarily blocked from content creation. Please retry your request again later.",
>   "documentation_url": "https://docs.github.com/rest/overview/resources-in-the-rest-api#secondary-rate-limits"
> }
```

The documentation says that we may take a look at `Retry-After` header. to know after what period to retry, but I think in our case this header was missing. 

Info about the rate limit properties we could get from the requests headers 
```
{
   "X-RateLimit-Limit":"5000",
   "X-RateLimit-Remaining":"0",
   "X-RateLimit-Reset":"1644427101",
   "X-RateLimit-Used":"5000",
   "X-RateLimit-Resource":"graphql",
}
```

So what we have discussed with Anna is that in our case we hit `403` we need to retry in a few seconds. 
If we hit the real rate limit (from 1. ), we need to take a look at  `X-RateLimit-Reset` header, and wait until this time - it could be just a few minutes or longer.
And dhe exception in 1. need to be first

difince reviewed (changes requested):
Let make the code handle the real rate limit exception. 
I would suggest doing a retry mechanism for the following error codes are well : 
```
RETRY_AFTER_STATUS_CODES = {frozenset: 3} frozenset({503, 413, 429})
 140601250491216 = {int} 503
 140601250490544 = {int} 413
 140601250490928 = {int} 429
```
to be sure that the retry mechanism will continue for 413 and 429 as well

annajung commented:
- Updated to handle codes: 502, 503, 403, 413, 429
- Added 2 seconds sleep before each call is made, verified that this resolves secondary rate limit
- Added support for real rate limit where status code is 200 but `X-RateLimit-Remaining` header returns 0
  - waits until rate limit resets based on `X-RateLimit-Reset`

PTAL! Thanks!

difince commented on ml-conversational-analytic-tool/githubDataExtraction.py:
```diff
@@ -1,179 +1,240 @@
+                        query = {'query': query['query'], 'variables': variables}
+
+                        data += nodes
+                elif status_code == 502 or status_code == 503 or status_code == 403 or status_code == 413 or status_code == 429:
```
nit: what about rewriting this line to `elis status_code in (...)` - to be shorter?

annajung replied:
Done

difince commented on ml-conversational-analytic-tool/githubDataExtraction.py:
```diff
@@ -1,179 +1,240 @@
+
+        while has_next_page:
+            # wait two seconds to avoid the secondary rate limits
+            time.sleep(2)
```
with my test, I rarely hit the secondary rate limit .. and adding 2 seconds on each iteration looks too much.

annajung replied:
I think the performance may depend on the machine of the user. Without any sleep, I run into 403 after a few calls have been made. Without waiting, it keeps making the call and does not resolve the 403 error.
```
 python ./ml-conversational-analytic-tool/githubDataExtraction.py kubernetes -R website  
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
```

I couldn't recreate the 403 error for when I set the wait time to 1 second, but I did run into it while testing yesterday. Therefore, decided to go with 2 seconds which I have not run into 403 error at all.

Since it's rarely thrown for 1 second, I'm happy to update to 1 second if 2 seconds is too long.

annajung replied:
UPDATE: I was able to recreate the 403 error with 1 second wait time using kubernetes/website 

```
retrieving data...
retrieving data...
retrieving data...
retrying due to 502 Bad Gateway {'data': None, 'errors': [{'message': 'Something went wrong while executing your query. This may be the result of a timeout, or it could be a GitHub bug. Please include `FCA9:69B8:55EC22:B9A6BF:620D4F00` when reporting this issue.'}]}
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrieving data...
retrieving data...
```

See that it runs into 403, but is able to resolve after waiting. Since it's pretty rare and it's able to resolve itself right away after waiting, I think updating to 1 second makes sense.

annajung replied:
Updated to 1 second

difince reviewed (approved):
LGTM
Thank you @annajung for implementing graphql functionality

annajung commented:
code update was to resolve merge conflicts caused by the merge of another PR https://github.com/vmware-labs/ml-conversational-analytic-tool/pull/36
